### PR TITLE
Native go building

### PR DESCRIPTION
**Pull request by fabxc**

### CODE REVIEW

**Comment by juliusv** [comment]
:+1: from me, but would be cool if @matthiasr could take a second look for the Makefile/shell stuff.

So `web/Makefile` stays for now?

**Comment by juliusv** [comment]
Actually, `web/Makefile` includes `Makefile.INCLUDE`, which is being removed here. So that needs fixing...

**Comment by matthiasr** [comment]
Simpler is better! Could you add a note to the README stating that Go 1.5.1 is required?

**Comment by juliusv** [comment]
Oh yeah, the README.md needs some updates. Like the sentence that goes "That includes a Go development environment of the appropriate version." etc.

**Comment by fabxc** [comment]
@juliusv ideas on the "no bash" error in circleci?

**Comment by juliusv** [comment]
@fabxc Add `bash` to the `apk add` line in the Dockerfile?

**Comment by fabxc** [comment]
Ah from Docker directly, thanks!

**Comment by juliusv** [comment]
@fabxc And the `mercurial bzr` deps can probably go away now, since we're not pulling any Go library dependencies over the network anymore?

**Comment by juliusv** [comment]
CircleCI still failing with:

```
can't load package: package github.com/prometheus/prometheus/cmd/prometheus: cannot find package "github.com/prometheus/prometheus/cmd/prometheus" in any of:
	/usr/lib/go/src/github.com/prometheus/prometheus/cmd/prometheus (from $GOROOT)
	($GOPATH not set)
```

`GOPATH` needs to be set, and for that to work, the root directory also needs to be copied into `/somethingsomething/src` instead of `/app`. Do you have Docker installed locally so you can test? Otherwise I can try to get it going.

The other question is how we ensure that it's using at least Go 1.5.1... maybe we need to download it ourselves.

**Comment by juliusv** [comment]
Ok yeah, I'm playing around with it locally, and the package that gets installed is Go 1.4.2, which doesn't support the new vendoring. I guess I have to add a manual download step...

**Comment by juliusv** [comment]
@fabxc Got it working like this: https://gist.github.com/juliusv/005eee74ab9a9577dc4d

File: Makefile [hunk 1]
```diff
@@ -11,86 +11,33 @@
...
+vet:
+	@echo ">> vetting code"
+	@$(GO) vet $(pkgs)
 
-binary: build
-
-build: dependencies $(GOPATH)
-	$(GO) build -o prometheus $(BUILDFLAGS) github.com/prometheus/prometheus/cmd/prometheus
-	$(GO) build -o promtool $(BUILDFLAGS) github.com/prometheus/prometheus/cmd/promtool
+build:
```
Comment (juliusv) [inline Makefile]: Trailing whitespace.

Comment (fabxc) [inline Makefile]: I was working on it but had to leave. Will update and look into yours
tomorrow morning.

**Comment by fabxc** [comment]
@juliusv finished what I started yesterday, which does the same as yours of course but keeps it a bit more general in the `scripts/goenv.sh` so it can be used elsewhere, too.

File: Dockerfile [hunk 2]
```diff
@@ -1,17 +1,18 @@
...
 MAINTAINER  The Prometheus Authors <prometheus-developers@googlegroups.com>
 
-WORKDIR /app
-COPY    . /app
+WORKDIR /gopath/src/github.com/prometheus/prometheus
+COPY    . /gopath/src/github.com/prometheus/prometheus
 
-RUN apk add --update -t build-deps git mercurial bzr make \
+RUN apk add --update -t build-deps tar openssl git make bash \
+    && source ./scripts/goenv.sh /go /gopath \
```
Comment (juliusv) [inline Dockerfile]: Let's also delete the `/go` directory at the end, together with the other dirs?

File: scripts/goenv.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	os=$(uname | tr A-Z a-z)
+	arch=$(uname -m | sed -e 's/x86_64/amd64/' | sed -e 's/i.86/386/')
+
+	url="https://golang.org/dl"
+	tarball="go${go_version_install}.${os}-${arch}.tar.gz"
+
+	wget -qO- "${url}/${tarball}" | tar -C "${GOROOT}" --strip 1 -xz
+fi
+
+PATH="$PATH:$GOROOT/bin"
```
Comment (juliusv) [inline scripts/goenv.sh]: Missing `export` here?

File: scripts/goenv.sh [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
```
Comment (juliusv) [inline scripts/goenv.sh]: nit: trailing empty line

File: scripts/goenv.sh (hunk 3 above)
Comment (fabxc) [inline scripts/goenv.sh]: Uh, script is meant to be sourced – but cannot hurt.

File: README.md [hunk 5]
```diff
@@ -33,44 +33,34 @@ is the recommended way of installing Prometheus.
...
+### Building from source
 
-Clone the repository in the usual way with `git clone`. (If you
-download and unpack the source archives provided by GitHub instead of
-using `git clone`, you need to set an environment variable `VERSION`
-to make the below work. See
-[issue #609](https://github.com/prometheus/prometheus/issues/609) for
-context.)
+To build Prometheus from the source code yourself you need to have a working
+go environment with [version 1.5 or greater installed](http://golang.org/doc/install).
```
Comment (juliusv) [inline README.md]: nit: "go environment" -> "Go environment"

File: README.md [hunk 6]
```diff
@@ -33,44 +33,34 @@ is the recommended way of installing Prometheus.
...
-development environment of the appropriate version.
+    $ mkdir -p $GOPATH/src/github.com/prometheus
+    $ cd $GOPATH/src/github.com/prometheus
+    $ git clone https://github.com/prometheus/prometheus.git
+    $ cd prometheus
+    $ make
+    $ ./prometheus -config.file=your_config.yml
 
-The `Makefile` offers a number of useful targets. Some examples:
+The makefile provides several targets:
```
Comment (juliusv) [inline README.md]: Makefile is still capitalized everywhere you look...

**Comment by juliusv** [comment]
:+1: otherwise

**Comment by fabxc** [comment]
All addressed :ship:

**Pull request merged**